### PR TITLE
[feature-wip](partial update) support MOW partial update in stream load

### PR DESCRIPTION
**Pull request by yixiutt**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.65 seconds
 stream load tsv:          450 seconds loaded 74807831229 Bytes, about 158 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          73 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230321201317_clickbench_pr_118130.html

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run feut

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 1]
```diff
@@ -297,11 +306,235 @@ void SegmentWriter::_maybe_invalid_row_cache(const std::string& key) {
...
+                                          _mow_context->max_version, &rowset);
+        if (st.is<NOT_FOUND>()) {
+            if (!_tablet_schema->allow_key_not_exist_in_partial_update()) {
+                return Status::InternalError("partial update key not exist before");
+            }
+            has_default = true;
+            use_default_flag.emplace_back(true);
+            continue;
+        }
+        if (!st.ok()) {
```
Comment (zhannngchen) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:411]: Should also handle `ALREADY_EXISTS` status?

File: be/src/olap/tablet.cpp [hunk 2]
```diff
@@ -2590,6 +2703,30 @@ Status Tablet::calc_delete_bitmap(RowsetId rowset_id,
...
+                        // version 4 when write, and when publish version, version 5's value will
+                        // be marked as deleted and it's update is losed.
+                        // So here we should read version 5's columns and build a new row, which is
+                        // consists of version 6's update columns and version 5's origin columns
+                        // here we build 2 read plan for ori values and update values
+                        prepare_to_read(loc, pos, &read_plan_ori);
+                        prepare_to_read(RowLocation {rowset_id, seg->id(), row_id}, pos,
+                                        &read_plan_update);
+                        rsid_to_rowset[rowset_find->rowset_id()] = rowset_find;
+                        ++pos;
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:2718]: We can't guarantee that the data written to rowset_writer is sorted, just by using the `pos` variable.

File: be/src/olap/tablet.cpp [hunk 3]
```diff
@@ -2601,6 +2738,9 @@ Status Tablet::calc_delete_bitmap(RowsetId rowset_id,
                     // just set 0 as a unified temporary version number, and update to
                     // the real version number later.
                     delete_bitmap->add({loc.rowset_id, loc.segment_id, 0}, loc.row_id);
+                    if (is_partial_update && rowset_writer != nullptr) {
+                        delete_bitmap->add({rowset_id, seg->id(), 0}, row_id);
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:2742]: duplicate delete might happen here? any problem?

File: be/src/olap/tablet.h [hunk 4]
```diff
@@ -403,20 +405,45 @@ class Tablet : public BaseTablet {
...
+    Status read_columns_by_plan(
+            TabletSchemaSPtr tablet_schema, const std::vector<uint32_t> cids_to_read,
+            const std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>& read_plan,
+            const std::map<RowsetId, RowsetSharedPtr>& rsid_to_rowset, vectorized::Block& block,
+            std::map<uint32_t, uint32_t>* read_index);
+    void prepare_to_read(const RowLocation& row_location, size_t pos,
+                         std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>* read_plan);
+    Status generate_new_block_for_partial_update(
+            TabletSchemaSPtr rowset_schema,
+            const std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>& read_plan_ori,
```
Comment (zhannngchen) [inline be/src/olap/tablet.h:438]: consider to use a type alias for `std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>>` to make the code clear, e.g.

typedef std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>> ReadPlan

**Review by zhannngchen — CHANGES_REQUESTED** [review]

File: be/src/olap/delta_writer.cpp [hunk 5]
```diff
@@ -269,10 +273,10 @@ void DeltaWriter::_reset_mem_table() {
         _mem_table_tracker.push_back(mem_table_insert_tracker);
         _mem_table_tracker.push_back(mem_table_flush_tracker);
     }
+    auto mow_context = std::make_shared<MowContext>(_cur_max_version, _rowset_ids, _delete_bitmap);
```
[the remark this answers is not in the document]
Comment (yixiutt) [reply be/src/olap/delta_writer.cpp:316]: 跟哪里重复了？

[the remark this answers is not in the document]
Comment (zhannngchen) [reply be/src/olap/delta_writer.cpp:316]: 忽略，这个是前面最开始review的时候留下来的一个想check的点，后来忘记删掉了

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 6]
```diff
@@ -297,11 +306,235 @@ void SegmentWriter::_maybe_invalid_row_cache(const std::string& key) {
...
+Status SegmentWriter::fill_missing_columns(vectorized::MutableColumns& mutable_full_columns,
+                                           const std::vector<bool>& use_default_flag,
+                                           bool has_default) {
+    // create old value columns
+    auto old_value_block = _tablet_schema->create_missing_columns_block();
+    std::vector<uint32_t> cids_missing = _tablet_schema->get_missing_cids();
+    CHECK(cids_missing.size() == old_value_block.columns());
+    auto mutable_old_columns = old_value_block.mutate_columns();
+    bool has_row_column = _tablet_schema->store_row_column();
+    // record real pos, key is input line num, value is old_block line num
```
[the remark this answers is not in the document]
Comment (yixiutt) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:459]: 这里面有些逻辑不一样，当时考虑了不好复用，fill_missing_column得考虑defaule value

File: be/src/olap/rowset/segment_v2/segment_writer.cpp (hunk 1 above)
Comment (yixiutt) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:411]: 这里应该不需要考虑，这里只是填充值，sequence列的处理逻辑应该跟之前是一样的

File: be/src/olap/tablet.cpp [hunk 7]
```diff
@@ -2515,15 +2611,29 @@ Status Tablet::_load_rowset_segments(const RowsetSharedPtr& rowset,
...
+    std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>> read_plan_ori;
+    std::map<RowsetId, std::map<uint32_t, std::vector<RidAndPos>>> read_plan_update;
+
+    std::map<RowsetId, RowsetSharedPtr> rsid_to_rowset;
+    rsid_to_rowset[rowset_id] = rowset;
+    vectorized::Block block = rowset_schema->create_block();
+    uint32_t pos = 0;
+    StorageReadOptions opt;
+    auto block_row_max = opt.block_row_max;
+
```
[the remark this answers is not in the document]
Comment (yixiutt) [reply be/src/olap/tablet.cpp:2636]: fixed

File: be/src/olap/tablet.cpp (hunk 2 above)
Comment (yixiutt) [reply be/src/olap/tablet.cpp:2718]: fixed，add a sort_block function to handle this

File: be/src/olap/tablet.h (hunk 4 above)
Comment (yixiutt) [reply be/src/olap/tablet.h:438]: fixed

**Comment by yixiutt** [comment]
run buildall

File: be/src/olap/tablet.cpp [hunk 8]
```diff
@@ -2582,19 +2741,48 @@ Status Tablet::calc_delete_bitmap(RowsetId rowset_id,
...
-
-                    // delete bitmap will be calculate when memtable flush and
-                    // publish. The two stages may see different versions.
-                    // When there is sequence column, the currently imported data
-                    // of rowset may be marked for deletion at memtablet flush or
-                    // publish because the seq column is smaller than the previous
-                    // rowset.
-                    // just set 0 as a unified temporary version number, and update to
-                    // the real version number later.
                     delete_bitmap->add({loc.rowset_id, loc.segment_id, 0}, loc.row_id);
```
Comment (zhannngchen) [inline be/src/olap/tablet.cpp:2786]: This line is useless

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Comment by yixiutt** [comment]
run buildall

**Review by zhannngchen — APPROVED** [review]
LGTM

**Pull request merged**